File: ios/IntervalTimer/Billing.swift
import Combine
import StoreKit
import SwiftUI

/// The one thing there is to buy. Must match the App Store Connect product ID exactly, and it is
/// deliberately the same string as the Play product ID — one name for one thing across both stores.
let UNLOCK_ID = "unlock"

/// Themes that never cost anything: the app has to be complete and good-looking without paying.
let FREE_PALETTES: Set<Palette> = [.standard, .mono]

/// Saved sequences on the free tier. Built-ins don't count — they aren't in presets.json.
let FREE_PRESET_SLOTS = 3

// The two gate rules as plain functions of (thing, entitlement), so they're testable without
// StoreKit, a sandbox account, or a running app. The Billing object below is only the entitlement.
func paletteLocked(_ p: Palette, unlocked: Bool) -> Bool { !unlocked && !FREE_PALETTES.contains(p) }
func presetsFull(_ savedCount: Int, unlocked: Bool) -> Bool { !unlocked && savedCount >= FREE_PRESET_SLOTS }

/// The $2.99 one-time unlock: six themes and unlimited saved sequences. Nothing about the timer
/// itself is behind it — not a phase, not a language, not background running.
///
/// The App Store is the source of truth, re-read on every launch, so "restore" is not something the
/// user has to do for the app to work — the button exists because Apple requires a visible restore
/// path for a non-consumable, and because a purchase made thirty seconds ago on another device
/// needs a way in. No receipt is cached anywhere it could be wrong.
///
/// Android's twin is `Billing.kt`, and the two are deliberately shaped the same. It has one thing
/// this doesn't: a grandfather flag for installs predating the paywall. iOS has never shipped, so
/// there is nobody to grandfather — the unlock is in version 1.0.0 from the first day, which is
/// also what Apple requires of a first non-consumable.
@MainActor
final class Billing: ObservableObject {
    static let shared = Billing()

    /// Bought it, on this Apple Account.
    @Published private(set) var purchased = false

    /// The App Store's own formatted price in the user's currency — never a hardcoded "$2.99".
    @Published private(set) var price: String?

    /// Set by any gate the user walks into; the root view shows the sheet when it's true.
    @Published var paywall = false

    /// What everything else asks. Named to match Android, where owning it and predating the paywall
    /// are two different things that mean the same to the app.
    var unlocked: Bool { purchased }

    private var product: Product?
    private var watcher: Task<Void, Never>?

    private init() {}

    /// Called once from the root view's `.task`. Reads what this account already owns, fetches the
    /// price, then keeps listening: `Transaction.updates` is how a purchase made on another device,
    /// an Ask-to-Buy approval, or a refund reaches a running app.
    func start() async {
        await refresh()
        watcher = Task { [weak self] in
            for await update in StoreKit.Transaction.updates {
                guard let self else { return }
                if case .verified(let t) = update {
                    await self.apply(t)
                    await t.finish()
                }
            }
        }
    }

    /// Re-ask the App Store what this account owns and what the unlock costs. Cheap, and the only
    /// thing standing between a reinstall and the themes the user already paid for.
    func refresh() async {
        var owns = false
        for await entitlement in StoreKit.Transaction.currentEntitlements {
            // `.unverified` is a transaction whose signature didn't check out. Ignoring it is the
            // whole point of verification — an unverified entitlement is not an entitlement.
            if case .verified(let t) = entitlement, t.productID == UNLOCK_ID, t.revocationDate == nil {
                owns = true
            }
        }
        purchased = owns
        product = try? await Product.products(for: [UNLOCK_ID]).first
        price = product?.displayPrice
    }

    /// Opens the App Store's own sheet. A no-op if the product hasn't loaded — no network, or the
    /// product isn't Ready to Submit in App Store Connect yet — so it re-fetches and lets the user
    /// tap again rather than showing an error for something a second tap usually fixes.
    func purchase() async {
        guard let product else { await refresh(); return }
        guard let result = try? await product.purchase() else { return }
        switch result {
        case .success(let verification):
            if case .verified(let t) = verification {
                await apply(t)
                // Unfinished transactions are replayed to the app forever. Finishing is what tells
                // StoreKit the goods were delivered.
                await t.finish()
            }
        // .pending is Ask to Buy waiting on a parent, and .userCancelled is a closed sheet. Neither
        // is an error, and neither grants anything — `Transaction.updates` delivers the approval if
        // and when it comes.
        case .pending, .userCancelled: break
        @unknown default: break
        }
    }

    /// Apple requires a visible restore control for a non-consumable. `AppStore.sync()` is the one
    /// that prompts for a password; `currentEntitlements` alone already covers the ordinary
    /// reinstall case on launch, which is why this is a button and not the mechanism.
    func restore() async {
        try? await AppStore.sync()
        await refresh()
    }

    private func apply(_ t: StoreKit.Transaction) async {
        guard t.productID == UNLOCK_ID else { return }
        // A refunded or revoked purchase arrives here too, carrying a revocationDate. Honouring it
        // is not optional: keeping the themes switched on after Apple has taken the money back is
        // the kind of thing that ends an account.
        purchased = t.revocationDate == nil
        if purchased { paywall = false }
    }
}

// MARK: - Paywall

/// The paywall. It appears only when the user has just reached for something behind it — a locked
/// theme, or a fourth saved sequence — never on launch, and never over a running workout.
///
/// Deliberately says what stays free before it says what costs money: the timer is the app, and
/// nobody should have to read to the bottom to find out the thing they came for still works.
struct UnlockSheet: View {
    @ObservedObject private var billing = Billing.shared
    let onDismiss: () -> Void

    var body: some View {
        VStack(spacing: 0) {
            Text("Unlock everything")
                .font(.system(size: 22, weight: .bold))
                .foregroundStyle(.white)
            Text("All eight themes.\nUnlimited saved sequences.")
                .font(.system(size: 16))
                .foregroundStyle(.white.opacity(0.85))
                .multilineTextAlignment(.center)
                .lineSpacing(4)
                .padding(.top, 14)
            Text("The timer, the cues, background running, all \(FREE_PRESET_SLOTS) free slots and "
                 + "every language stay free forever. One payment, no subscription.")
                .font(.system(size: 13))
                .foregroundStyle(.white.opacity(0.45))
                .multilineTextAlignment(.center)
                .lineSpacing(3)
                .padding(.top, 14)

            // The App Store's formatted price, in the user's own currency. Until it arrives the
            // button says what it does and nothing about money — a hardcoded "$2.99" would be a lie
            // in most of the places this app ships.
            GlassPill(
                text: billing.price.map { "Unlock — \($0)" } ?? "Unlock",
                action: { Task { await billing.purchase() } },
                wide: true,
                big: true
            )
            .padding(.top, 22)

            Button("Restore purchase") { Task { await billing.restore() } }
                .font(.system(size: 14))
                .foregroundStyle(.white.opacity(0.5))
                .buttonStyle(.plain)
                .padding(.top, 14)

            Button("Not now", action: onDismiss)
                .font(.system(size: 15))
                .foregroundStyle(.white.opacity(0.5))
                .buttonStyle(.plain)
                .padding(.top, 12)
        }
        .padding(24)
        .frame(maxWidth: .infinity)
        // Solid black under the glass: this is presented over the app, and the translucent fill
        // alone would read as a smear of the home screen rather than as a surface.
        .background(Color.black)
        .background(glassFill)
        .presentationDetents([.height(430)])
        .presentationBackground(.black)
        .preferredColorScheme(.dark)
    }
}

File: app/src/main/java/com/chrispoole/intervaltimer/Billing.kt
package com.chrispoole.intervaltimer

import android.app.Activity
import android.content.Context
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.core.content.edit
import com.android.billingclient.api.AcknowledgePurchaseParams
import com.android.billingclient.api.BillingClient
import com.android.billingclient.api.BillingClientStateListener
import com.android.billingclient.api.BillingFlowParams
import com.android.billingclient.api.BillingResult
import com.android.billingclient.api.PendingPurchasesParams
import com.android.billingclient.api.ProductDetails
import com.android.billingclient.api.Purchase
import com.android.billingclient.api.QueryProductDetailsParams
import com.android.billingclient.api.QueryPurchasesParams
import com.chrispoole.intervaltimer.ui.Palette

/** The one thing there is to buy. Must match the Play Console product ID exactly. */
const val UNLOCK_SKU = "unlock"

/** Themes that never cost anything: the app has to be complete and good-looking without paying. */
val FREE_PALETTES = setOf(Palette.DEFAULT, Palette.MONO)

/** Saved sequences on the free tier. Built-ins don't count — they aren't in presets.json. */
const val FREE_PRESET_SLOTS = 3

// The two gate rules as plain functions of (thing, entitlement), so they're testable without a
// BillingClient, a Play Store, or an Activity. The Billing object below is only the entitlement.
fun paletteLocked(p: Palette, unlocked: Boolean) = !unlocked && p !in FREE_PALETTES
fun presetsFull(savedCount: Int, unlocked: Boolean) = !unlocked && savedCount >= FREE_PRESET_SLOTS

/**
 * The $2.99 one-time unlock: six themes and unlimited saved sequences. Nothing about the timer
 * itself is behind it — not a phase, not a language, not background running.
 *
 * Play is the source of truth, re-queried on every launch, so "restore" is not a thing the user has
 * to do — the row exists anyway for the case where they bought it thirty seconds ago on another
 * device. No receipt is cached beyond [purchased], because a cache is a thing that can be wrong and
 * queryPurchasesAsync is free.
 */
object Billing {
    private var client: BillingClient? = null
    private var details: ProductDetails? = null

    /** Bought it, on this Play account. */
    var purchased by mutableStateOf(false); private set

    /**
     * Installed before the unlock existed. These people already have all eight themes and as many
     * presets as they made; taking them away in an update is how you earn a one-star review from
     * someone who liked the app. Decided once, on the first launch of the version that added this,
     * and written down — `firstInstallTime < lastUpdateTime` means "this install was updated into
     * the paywall", but it stays true for every later update too, so it must not be re-asked.
     */
    private var legacy = false

    /** What everything else asks. Owning it and predating it are the same thing to the app. */
    val unlocked: Boolean get() = purchased || legacy

    /** Play's own formatted price in the user's currency — never a hardcoded "$2.99" in the UI. */
    var price by mutableStateOf<String?>(null); private set

    /** Set by any gate the user walks into; the root composable shows the sheet when it's true. */
    var paywall by mutableStateOf(false)

    fun init(context: Context) {
        if (client != null) return
        val app = context.applicationContext
        val prefs = app.getSharedPreferences("billing", Context.MODE_PRIVATE)
        if (!prefs.contains("legacy")) {
            val pi = app.packageManager.getPackageInfo(app.packageName, 0)
            prefs.edit { putBoolean("legacy", pi.firstInstallTime < pi.lastUpdateTime) }
        }
        legacy = prefs.getBoolean("legacy", false)

        client = BillingClient.newBuilder(app)
            .setListener { result, purchases -> onPurchases(result, purchases) }
            .enablePendingPurchases(PendingPurchasesParams.newBuilder().enableOneTimeProducts().build())
            .build()
        connect()
    }

    private fun connect() {
        val c = client ?: return
        if (c.isReady) { refresh(); return }
        c.startConnection(object : BillingClientStateListener {
            override fun onBillingSetupFinished(result: BillingResult) {
                if (result.responseCode == BillingClient.BillingResponseCode.OK) refresh()
            }
            // No retry loop. Nothing is broken while billing is down — the app is a timer, and the
            // next launch (or the next tap on Unlock) reconnects.
            override fun onBillingServiceDisconnected() {}
        })
    }

    /** Re-ask Play what this account owns and what the unlock costs. Cheap; called on every launch. */
    fun refresh() {
        val c = client?.takeIf { it.isReady } ?: run { connect(); return }
        c.queryPurchasesAsync(
            QueryPurchasesParams.newBuilder().setProductType(BillingClient.ProductType.INAPP).build(),
        ) { _, purchases -> onPurchases(null, purchases) }

        c.queryProductDetailsAsync(
            QueryProductDetailsParams.newBuilder().setProductList(
                listOf(
                    QueryProductDetailsParams.Product.newBuilder()
                        .setProductId(UNLOCK_SKU)
                        .setProductType(BillingClient.ProductType.INAPP)
                        .build(),
                ),
            ).build(),
        ) { _, result ->
            details = result.productDetailsList.firstOrNull()
            price = details?.oneTimePurchaseOfferDetails?.formattedPrice
        }
    }

    /** Opens Play's own sheet. Silently a no-op if billing never connected — see [connect]. */
    fun purchase(activity: Activity) {
        val c = client?.takeIf { it.isReady } ?: run { connect(); return }
        val pd = details ?: run { refresh(); return }
        c.launchBillingFlow(
            activity,
            BillingFlowParams.newBuilder().setProductDetailsParamsList(
                listOf(BillingFlowParams.ProductDetailsParams.newBuilder().setProductDetails(pd).build()),
            ).build(),
        )
    }

    /**
     * One path for both a fresh purchase and a launch-time restore, so the two can't disagree about
     * what counts as owned. PENDING (cash at a kiosk, a parent's approval) deliberately grants
     * nothing yet — it isn't paid — and resolves on a later launch's query.
     */
    private fun onPurchases(result: BillingResult?, purchases: List<Purchase>?) {
        if (result != null && result.responseCode != BillingClient.BillingResponseCode.OK) return
        val owned = purchases.orEmpty().filter {
            it.purchaseState == Purchase.PurchaseState.PURCHASED && UNLOCK_SKU in it.products
        }
        if (owned.isEmpty()) return
        purchased = true
        paywall = false
        // Unacknowledged for three days and Play refunds it automatically, so this is the whole
        // difference between being paid and not.
        owned.filterNot { it.isAcknowledged }.forEach { p ->
            client?.acknowledgePurchase(
                AcknowledgePurchaseParams.newBuilder().setPurchaseToken(p.purchaseToken).build(),
            ) { }
        }
    }
}
